resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/KerberosConfDriverFeatureStep.scala

```diff
@@ -0,0 +1,161 @@
+        " you must also specify the name of the secret",
+      "If a secret storing a Kerberos Delegation Token is specified you must also" +
+        " specify the item-key where the data is stored")
+
+    require(kubernetesConf.hadoopConfDir.isDefined, "Ensure that HADOOP_CONF_DIR is defined")
+    private val hadoopConfDir = kubernetesConf.hadoopConfDir.get
+    private val hadoopConfigurationFiles = HadoopBootstrapUtil.getHadoopConfFiles(hadoopConfDir)
+
+    // Either use pre-existing secret or login to create new Secret with DT stored within
+    private val hadoopSpec: Option[KerberosConfigSpec] = (for {
+      secretName <- maybeExistingSecretName
+      secretItemKey <- maybeExistingSecretItemKey
+    } yield {
+      KerberosConfigSpec(
+         dtSecret = None,
+         dtSecretName = secretName,
+         dtSecretItemKey = secretItemKey,
+         jobUserName = kubeTokenManager.getCurrentUser.getShortUserName)
+    }).orElse(
+      if (isKerberosEnabled) {
+         Some(HadoopKerberosLogin.buildSpec(
+             conf,
+             kubernetesConf.appResourceNamePrefix,
+             kubeTokenManager))
+       } else None )
+
+    override def configurePod(pod: SparkPod): SparkPod = {
+      val hadoopBasedSparkPod = HadoopBootstrapUtil.bootstrapHadoopConfDir(
+        hadoopConfDir,
+        kubernetesConf.hadoopConfigMapName,
+        pod)
+      hadoopSpec.map { hSpec =>
+        HadoopBootstrapUtil.bootstrapKerberosPod(
+            hSpec.dtSecretName,
+            hSpec.dtSecretItemKey,
+            hSpec.jobUserName,
+            maybeKrb5File,
+            kubernetesConf.kRBConfigMapName,
+            maybeKrb5CMap,
+            hadoopBasedSparkPod)
+      }.getOrElse(
+        HadoopBootstrapUtil.bootstrapSparkUserPod(
+          kubeTokenManager.getCurrentUser.getShortUserName,
+          hadoopBasedSparkPod))
+    }
+
+    override def getAdditionalPodSystemProperties(): Map[String, String] = {
+      val resolvedConfValues = hadoopSpec.map{ hSpec =>
+         Map(KERBEROS_KEYTAB_SECRET_NAME -> hSpec.dtSecretName,
+            KERBEROS_KEYTAB_SECRET_KEY -> hSpec.dtSecretItemKey,
+            KERBEROS_SPARK_USER_NAME -> hSpec.jobUserName)
+      }.getOrElse(
+          Map(KERBEROS_SPARK_USER_NAME ->
+            kubeTokenManager.getCurrentUser.getShortUserName))
+      Map(HADOOP_CONFIG_MAP_NAME -> kubernetesConf.hadoopConfigMapName,
+          HADOOP_CONF_DIR_LOC -> kubernetesConf.hadoopConfDir.get) ++ resolvedConfValues
+    }
+
+    override def getAdditionalKubernetesResources(): Seq[HasMetadata] = {
+      val krb5ConfigMap = maybeKrb5File.map { fileLocation =>
+          HadoopBootstrapUtil.buildkrb5ConfigMap(
+          kubernetesConf.kRBConfigMapName,
+          fileLocation)}
+      val kerberosDTSecret = for {
+        hSpec <- hadoopSpec
+        kDtSecret <- hSpec.dtSecret
+      } yield {
+        kDtSecret
+      }
+      val configMap =
```

Comment:
Can you move this into a utility method maybe named `buildHadoopConfigMap` into `HadoopBoostrapUtil`?